src/core/bpf-lsm.c

```diff
@@ -0,0 +1,264 @@
+
+        assert(filesystems);
+        assert(cgroup_path);
+
+        r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, cgroup_path, NULL, &path);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get systemd cgroup path");
+
+        r = cg_path_get_cgroupid(path, &cgroup_id);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get cgroup ID for path '%s'", path);
+
+        inner_map_fd = bpf_create_map(
+            BPF_MAP_TYPE_HASH,
+            sizeof(uint32_t),
+            sizeof(uint32_t),
+            128, /* Should be enough for all filesystem types */
+            0);
+        if (inner_map_fd < 0)
+                return log_error_errno(errno, "Failed to create inner LSM map: %m");
+
+        outer_map_fd = bpf_obj_get(LSM_MAP_BPFFS_PATH);
+        if (outer_map_fd < 0)
+                return log_error_errno(errno, "Error getting pinned LSM BPF map: %m");
+
+        if (bpf_map_update_elem(outer_map_fd, &cgroup_id, &inner_map_fd, BPF_ANY) != 0)
+                return log_error_errno(errno, "Error populating LSM BPF map: %m");
+
+        STRV_FOREACH(f, filesystems) {
+                if (DEBUG_LOGGING)
+                        log_debug("Restricting filesystem access to: %s", *f);
+
+                r = fs_type_from_string(*f, &magic);
+                if (r == -EINVAL) {
+                        log_warning("Unknown filesystem '%s'. Ignoring.", *f);
+                        continue;
+                }
+                if (r < 0)
+                        return r;
+
+                if (bpf_map_update_elem(inner_map_fd, &magic, &dummy_value, BPF_ANY) != 0) {
+                        r = log_error_errno(errno, "Failed to update BPF map: %m");
+
+                        if (bpf_map_delete_elem(outer_map_fd, &cgroup_id) != 0)
+                                log_debug_errno(errno, "Failed to delete cgroup entry from LSM BPF map: %m");
+
+                        return r;
+                }
+        }
+
+        return 0;
+}
+
+int cleanup_lsm_bpf(const char *cgroup_path) {
+        _cleanup_close_ int fd = -1;
+        _cleanup_free_ char *path = NULL;
+        uint64_t cgroup_id;
+        int r;
+
+        assert(cgroup_path);
+
+        if (!lsm_bpf_supported())
+                return 0;
+
+        r = cg_get_path(SYSTEMD_CGROUP_CONTROLLER, cgroup_path, NULL, &path);
+        if (r < 0)
+                return r;
+
+        r = cg_path_get_cgroupid(path, &cgroup_id);
+        if (r < 0)
+                return r;
+
+        fd = bpf_obj_get(LSM_MAP_BPFFS_PATH);
+        if (fd < 0)
+                return log_debug_errno(errno, "Failed to get LSM Map object from bpffs: %m");
+
+        if (bpf_map_delete_elem(fd, &cgroup_id) != 0)
+                return log_debug_errno(errno, "Failed to delete cgroup entry from LSM BPF map: %m");
+
+        return 0;
+}
+#else /* ! BUILD_BPF */
+int lsm_bpf_supported(void) {
+        return 0;
+}
+
+int lsm_bpf_setup(void) {
+        return -EOPNOTSUPP;
```

Comment:
this should probably output an error at LOG_ERR level.

Our coding style regarding error logging is that either a function is of the logging kind or of the non-logging kind. The latter are "library-like" functions where logging is left for the caller to do, the latter are more "main-program-like" functions, where logging is done itself, and the caller shouldn't log anymore. You should decide what kind of function your is (I'd say: main-program-like) and then make sure you log in all cases or in no cases. But in this current version of your patch you log in most error cases at LOG_ERR level, but if supported is not compiled in then you will not log at all...

So, the goal really is to avoid duplicate logging as well as missing logging: we want that every error is logged exactly once, to make things as little confusing as possible.

(Note that in the discussion above LOG_DEBUG level logging doesn't count as real logging... debug logging is always OK, regardless if library-like or main-program-like)